src/protocol/modulus_conversion/convert_shares.rs

```diff
@@ -134,28 +150,47 @@ where
 /// Propagates errors from convert shares
 /// # Panics
 /// Propagates panics from convert shares
-pub async fn convert_all_bits<F, C, S>(
+pub async fn convert_all_bits<'a, F, C, S, I>(
     ctx: &C,
-    locally_converted_bits: &[Vec<BitConversionTriple<S>>],
-) -> Result<Vec<Vec<S>>, Error>
+    mut locally_converted_bits: I,
+    num_bits: u32,
+    num_multi_bits: u32,
+) -> Result<impl Iterator<Item = Vec<Vec<S>>> + 'a, Error>
 where
     F: Field,
     C: Context<F, Share = S>,
-    S: ArithmeticSecretSharing<F>,
+    S: ArithmeticSecretSharing<F> + 'a,
+    I: Iterator<Item = Vec<BitConversionTriple<S>>>,
 {
-    let futures = locally_converted_bits
-        .iter()
+    let num_records = locally_converted_bits.size_hint().0;
+
+    let count = ((num_bits + num_multi_bits - 1) / num_multi_bits) as usize;
+    let mut futures = Vec::with_capacity(count);
+    futures.resize_with(count, || Vec::with_capacity(num_records));
+
+    locally_converted_bits
+        .by_ref()
         .enumerate()
-        .map(|(bit_num, one_column)| {
-            convert_bit_list(
-                ctx.narrow(&ModulusConversion(bit_num.try_into().unwrap()))
-                    .set_total_records(one_column.len()),
-                one_column,
-            )
-        })
-        .collect::<Vec<_>>();
-    let converted_shares = try_join_all(futures).await?;
-    Ok(converted_shares)
+        .for_each(|(idx, record)| {
+            (0..count).for_each(|slice_num| {
+                let first_bit =
+                    num_multi_bits * <usize as TryInto<u32>>::try_into(slice_num).unwrap();
```

Comment:
I think you can do this in a much more idiomatic way with `.chunks(num_multi_bits)`.

https://docs.rs/itertools/0.10.5/itertools/trait.Itertools.html#method.chunks